src/nupic/research/connections.py

```diff
@@ -52,92 +162,181 @@ def __init__(self,
+    return [synapse
+            for synapse in self._synapsesForSegment[segmentData.flatIdx]
+            if not synapse.data.destroyed]
 
 
   def dataForSynapse(self, synapse):
     """ Returns the data for a synapse.
 
-    @param synapse (int) Synapse index
+    @param synapse (Object) Synapse object
 
     @return (SynapseData) Synapse data
     """
 
-    return self._synapses[synapse]
+    return synapse.data
 
 
-  def synapsesForSegment(self, segment):
-    """ Returns the synapses on a segment.
+  def dataForSegment(self, segment):
+    """ Returns the data for a segment.
 
-    @param segment (int) Segment index
+    @param segment (Object) Segment object
+
+    @return (SegmentData) Segment data
+    """
+
+    return segment.data
+
+
+  def retrieveSegmentData(self, idx, cell):
+    """ Returns the data for a segment stored in the self._cells array
+
+    @param idx  (int) segment index on a cell
+    @param cell (int) cell index
+
+    @return (SegmentData) Segment data
 
-    @return (set) Synapse indices
+    NOTE: Only used in the read method in temporal memory for re-creating the
+          active and matching segments
     """
-    self._validateSegment(segment)
 
-    if not segment in self._synapsesForSegment:
-      return set()
+    return self._cells[cell].segments[idx]
+
+
+  def _leastRecentlyUsedSegment(self, cell):
+    """ Internal method for finding the least recently activated segment on a
+        cell
+
+    @param cell (int) cell index to search for segments on
+
+    @return (Object) Segment object of the segment that was least recently
+                     created/activated.
+    """
+    segments = self._cells[cell].segments
+    minIdx = float("inf")
+    minIteration = float("inf")
+
+    for i in xrange(len(segments)):
+      segment = segments[i]
+      if (not segment.destroyed and
+          segment.lastUsedIteration < minIteration):
+        minIdx = i
+        minIteration = segment.lastUsedIteration
+
+    return Segment(minIdx, cell, segments[minIdx])
+
 
-    return self._synapsesForSegment[segment]
+  def _minPermanenceSynapse(self, segment):
+    """ Internal method for finding the synapse with the smallest permanence
+    on the given segment.
+
+    @param segment (Object) Segment object to search for synapses on
+
+    @return (Object) Synapse object on the segment of the minimal permanence
+
+    Note: On ties it will chose the first occurance of the minimum permanence
+    """
+    synapses = self._synapsesForSegment[segment.data.flatIdx]
+    minIdx = float("inf")
+    minPermanence = float("inf")
+
+    for i in xrange(len(synapses)):
+      synapseData = synapses[i].data
+      if (not synapseData.destroyed) and (synapseData.permanence
+                                          < minPermanence - EPSILON):
+        minIdx = i
+        minPermanence = synapseData.permanence
+
+    return synapses[minIdx]
+
+
+  def mostActiveSegmentForCells(self, cells, inputCells, synapseThreshold):
```

Comment:
I think this method is dead code. We can remove it and the tests...

I removed them in the C++ here: https://github.com/numenta/nupic.core/pull/998#discussion_r70546162